tests: integration: self_update_test: add self-update test

aquilamacedo opened the pull request:
Introduce integration tests for self-update with
the --unstable option and regular self-update.

aquilamacedo commented:
TODO:

- [ ] Hide the output (it's printing the entire self-update process).

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,61 @@
+{
+  self_update_test_helper 'unstable'
+}
+
```
One test that I want to see:
1. Manually install kw in the container. Check if everything is all right.
2. Run an update to master (do the same for the unstable) and check if everything looks all right.

Do a similar test, but in this case, intentionally install an old version of the kw (maybe three commits old, you can force it via git) and ensure the update works as expected.

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,61 @@
+  expected_output=$(printf '%s\nBranch: %s\nCommit: %s' "$base_version" "$kw_branch" "$head_commit_hash")
+
+  for distro in "${DISTROS[@]}"; do
+    say "Running tests on ${distro} environment..."
```
I think this is unnecessary as it is outputted whether the test fails or succeeds. IMHO, the distro info is only relevant if the test fails (which is displayed by the assertion).

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,61 @@
+
+    case "$update_type" in
+      'master')
+        container_exec "${container}" 'yes | kw self-update'
```
Also, redirect the output of this command to `/dev/null` so as not to pollute the test suite output.

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,61 @@
+        container_exec "${container}" 'yes | kw self-update'
+        ;;
+      'unstable')
+        container_exec "${container}" 'yes | kw self-update --unstable'
```
Also, redirect the output of this command to `/dev/null` so as not to pollute the test suite output.

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,61 @@
+  done
+}
+
+function test_self_update_master_helper()
```
Remove the suffix `_helper` here and in the next test function.

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,61 @@
+  local head_commit_hash
+
+  # Get the short commit hash of the branch
+  head_commit_hash=$(git rev-parse --short origin/"${kw_branch}")
```
Also, this git command is run in the host repository (i.e., the developer's actual kw repo), in which we can't assume too much. For example, my kw fork is labeled `davidbtadokoro-fork` and the kw upstream repo `upstream`, so when I run the tests, I encounter the error `fatal: Needed a single revision` and the `Commit` field of the expected output gets blank in turn (see screenshot below).

![Screenshot from 2024-03-10 19-18-31](https://github.com/kworkflow/kworkflow/assets/95055784/cc6da6cd-acd4-4e56-ae93-3c0094d74362)

**Note:** The test output differs from this PR version because I cleaned the extra output and used `assert_equals_helper` for better visualization.

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,61 @@
+  head_commit_hash=$(git rev-parse --short origin/"${kw_branch}")
+
+  # Get the base version from the specified branch
+  base_version=$(git show "${kw_branch}":./src/VERSION | head --lines 1)
```
Like in the git command above, we assume some things about the dev's local kw repo. For example, I didn't have the `master` branch update with the upstream, and `src/VERSION` still contains `alpha` as the base version in this local branch. This causes the expected base version to be wrong (see screenshot below).

![Screenshot from 2024-03-10 19-23-35](https://github.com/kworkflow/kworkflow/assets/95055784/a31402f9-1f19-48da-8ffb-1b1300b13415)

**Note:** As in the previous screenshot, the output differs because I cleaned the extra output and used `assert_equals_helper`. Also, I "fixed" the problem pointed in the screenshot to make sure that the test failed strictly due to the expected base version being wrong.

davidbtadokoro reviewed (changes requested):
Hi @aquilamacedo, and thank you so much for this change!

I left some comments, so please see if you can address them (ping me if you have any questions). The most critical comments are about the verbose output that pollutes the test suite and the weak assumptions that the git commands make (the ones with screenshots).

**UPDATE:** I see now that you made a task to hide the output. Nevertheless, don't forget to do it!

aquilamacedo commented:
Hi @davidbtadokoro. Could you check if everything is in order?

I addressed all your suggestions, thx :)

Now I'm working on @rodrigosiqueira  suggestions

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,66 @@
+  local expected_output
+  local head_commit_hash
+
+  tmp_kw_dir=$(mktemp --directory)
```
I'm kind of confused: shouldn't this command be run inside the container? I think the temporary directory is being created on the host machine and that the test doesn't fail because the path returned will be something like `/tmp/tmp.XXXXXX`, which is a valid path for the container. Confirm if this is the case and, if so, run this command inside each container with `container_exec`.

aquilamacedo replied:
Hi @davidbtadokoro.

I hadn't noticed that, it's been fixed now. Thank you :-)

davidbtadokoro commented:
Hey, @aquilamacedo, and thank you for tackling my comments :smile: 

I ran the tests on my local machine, and it works. I think you solved the problem of the environment assumptions I mentioned. I just had a simple comment about creating the temp dir inside the host rather than inside the containers, so please address it.

Apart from that, everything LGTM!

rodrigosiqueira commented:
@aquilamacedo, any updates here?

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,141 @@
+include './tests/unit/utils.sh'
+include './tests/integration/utils.sh'
+
+declare -g expected_output
```
Do we need this global variable? FWIU, `expected_output` is only used inside `self_update_and_check_version`, so can't we use a local variable for this?

davidbtadokoro commented on tests/integration/utils.sh:
```diff
@@ -219,3 +219,14 @@ function container_inspect()
+  container_exec "$container" 'yes | ./setup.sh --completely-remove > /dev/null'
+
+  if [ "$?" -ne 0 ]; then
+    fail "Failed to completely remove kworkflow"
```
Use single quotes here

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,141 @@
+
+  output=$(container_exec "$container" "kw --version")
+
+  assert_equals_helper "kw self-update ${kw_branch} failed for ${distro}" "$LINENO" "$expected_output" "$output"
```
I really like the modular approach you've taken here, as the setup and the self-update vary across the cases in just some parameters. But can't we move the assertions (this line and the line where we get `output`) to the callers, i.e., the functions that represent the test cases? I say this because the way the test for the feature is implemented, we lose the information `LINENO` (all will indicate fail in line 89) and a more fine-grained error message.

Sorry for being too demanding, but as test cases grow, these two pieces of information will matter when debugging failing tests.

rodrigosiqueira replied:
I agree

aquilamacedo replied:
I agree too, in this regard of pointing to the error line it would be more difficult to debug, i've already fixed this

davidbtadokoro reviewed (changes requested):
Hi @aquilamacedo, and thanks again for the update (self-update :thinking:)

This PR is legitimate, and I just left some comments that I think will be simple for you to tackle. The test cases are aligned with what @rodrigosiqueira asked, and they are running well, at least on my end and in the CI.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,141 @@
+
+  container_exec "$container" "git clone ${kw_repo_url} ${kw_tmp_dir}"
+  if [[ "$?" -ne 0 ]]; then
+    complain 'Git clone operation failed'
```
Add another `complain` highlighting the failed command, i.e., `git clone ${kw_repo_url} ${kw_tmp_dir}.
Do it for all of these types of checks.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,141 @@
+      ;;
+  esac
+
+  output=$(container_exec "$container" "kw --version")
```
Use single quotes for the kw --version

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,141 @@
+    setup_kworkflow_environment "$container" "$kw_tmp_dir" 'unstable' 'HEAD~4'
+
+    for kw_branch in "${kw_branches[@]}"; do
+      self_update_and_check_version "$container" "$kw_tmp_dir" "$kw_branch" "$distro"
```
It will be more readable if the assert is here.

rodrigosiqueira commented:
@aquilamacedo , could you check the CI failure?

aquilamacedo commented:
> @aquilamacedo , could you check the CI failure?

Yeah, I'm looking into that rn

aquilamacedo commented:
Since this PR was dependent on another one that has already been accepted, I have just updated it :)

aquilamacedo commented:
I made all the changes we discussed in the call, could you check them? @rodrigosiqueira @davidbtadokoro

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,213 @@
+{
+  local kw_repo_url='https://github.com/kworkflow/kworkflow'
+
+  HOST_KW_CLONED_PATH="$(mktemp --directory)/kworkflow"
```
I think this is a nitpick, but maybe we want to create another readonly global variable to store the path returned by `mktemp --directory`, so we can remove this temporary dir in `oneTimeTearDown` or below, in case of fail. At the moment, we are only removing the `kworkflow` repo inside the temp dir (again, a nitpick, as this temp dir wont occupy practically any space...)

aquilamacedo replied:
Nice catch, I added this change :)

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,213 @@
+# @kw_cloned_container_path: Temporary directory inside the container where kworkflow is cloned
+# @kw_branch: The branch of kworkflow to be used.
+# @commit_point: The specific commit to be checked out.
+function setup_outdated_kworkflow_environment()
```
Just to be sure, the idea is to:

1) in each `_old_version_` test case, install an arbitrary outdated upstream version of kw, using the cloned repo
2) in each `_old_version_` test case, get the expected version from `kw --version` using the the cloned repo
3) in each `_old_version_` test case, run the current version of `kw self-update` (from the local repo actually being tested.

If so, I feel like it is a great approach to circumvent some of the difficulties we mentioned in the offline conversation.

aquilamacedo replied:
Exactly, that's the idea.

davidbtadokoro commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,213 @@
+
+# This function tests the self-update functionality of kworkflow for the master
+# branch.
+function test_self_update_master()
```
Maybe I am missing something, but in this test case, and the below one, the kw version installed before the `container_exec "$container" 'yes | bash ./kw self-update > /dev/null'` line, is the version of the current repository that is being tested?

aquilamacedo replied:
Yes, before calling these functions, the version of `kw` matches the current version modified by the developer on the host.

davidbtadokoro reviewed (changes requested):
Hi @aquilamacedo, thanks a lot for the great update and sorry for the **massive** delay (on all PRs).

This PR is looking really great and refined. I left really small comments (some are just questions). Tested it locally and it works really nice.

The only "serious" issue is the execution time that it is really slow. I suppose there is no trivial way to solve this due to the fact that we need to clone instances of the kw repo to isolate the test git commands and don't mess with the current local repo. The `kw self-update` feature itself clones a repo, so I think there isn't much where to run...

Overall, wonderful work!

aquilamacedo commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,213 @@
+include './tests/integration/utils.sh'
+
+declare -a KW_BRANCHES=('master' 'unstable')
+declare -g HOST_KW_CLONED_PATH
```
`KW_BRANCHES` can be read-only with -ar. However, `HOST_KW_CLONED_PATH` needs to be assigned a value later, so it can't be read-only (-gr).

Also, combining declaration and assignment triggers a shellcheck `SC2155` error.

aquilamacedo commented:
> Hi @aquilamacedo, thanks a lot for the great update and sorry for the **massive** delay (on all PRs).
> 
> This PR is looking really great and refined. I left really small comments (some are just questions). Tested it locally and it works really nice.
> 
> The only "serious" issue is the execution time that it is really slow. I suppose there is no trivial way to solve this due to the fact that we need to clone instances of the kw repo to isolate the test git commands and don't mess with the current local repo. The `kw self-update` feature itself clones a repo, so I think there isn't much where to run...
> 
> Overall, wonderful work!

Thx for the review, David :)

I believe I have addressed all your comments. Could you please check?

Regarding the execution time, it is indeed a bit slow. I believe it is taking a long time because of the installations and clones of the kw repository that `self-update` does, as you said

rodrigosiqueira commented:
Hi @aquilamacedo ,
When I tried to run it in my local, I saw this error:

```bash
./run_tests.sh -v --integration test self_update_test 
Running test [integration/self_update_test]
=========================================================
Preparing environment for kw self-update integration tests...
[1/6] Using cached container image for archlinux.
[2/6] Creating archlinux container.
[3/6] Using cached container image for debian.
[4/6] Creating debian container.
[5/6] Using cached container image for fedora.
[6/6] Creating fedora container.
test_self_update_master
test_self_update_unstable
test_self_update_with_kw_old_version_master
ASSERT:(416): Failed to execute the command in the container.
kw setup failed
./setup.sh --install --force --skip-checks --skip-docs
test_self_update_with_kw_old_version_unstable

Ran 4 tests.

FAILED (failures=1)

Tearing down containers used in integration tests...
[1/3] Removing archlinux container.
[2/3] Removing debian container.
[3/3] Removing fedora container.
```

Any idea of what could be the issue?

@davidbtadokoro , could you also try on your side?

rodrigosiqueira commented:
@davidbtadokoro did you see my previous comment?

aquilamacedo commented:
Hi @rodrigosiqueira.

Try to clean your environment first:

 `
./run_tests.sh --integration clear-cache
`

Then, execute the self-update test with this command:

`
  ./run_tests.sh -v --integration test self_update_test 
 `

davidbtadokoro commented:
> @davidbtadokoro did you see my previous comment?

Hi, really sorry for the delay. I've tried on my end and couldn't reproduce your errors. I tested on a new environment where I hadn't ever run the integration tests, so (based on Aquila's response above) maybe this is why...

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+# @container: The name of the container.
+# @kw_cloned_container_path: Temporary directory inside the container where kworkflow is cloned
+# @kw_branch: The branch of kworkflow to be used.
+function get_expected_output()
```
This name is very generic; use a name that better describes what the expected output is.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+#
+# @container: The name of the container.
+# @kw_cloned_container_path: Temporary directory inside the container where kworkflow is cloned
+# @kw_branch: The branch of kworkflow to be used.
```
Document the return part of this function.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+  head_commit_hash=$(container_exec "$container" "git -C ${kw_cloned_container_path} rev-parse --short origin/${kw_branch}")
+  base_version=$(container_exec "$container" "git -C ${kw_cloned_container_path} show remotes/origin/${kw_branch}:./src/VERSION | head --lines 1")
+
+  expected_output=$(printf '%s\nBranch: %s\nCommit: %s' "$base_version" "$kw_branch" "$head_commit_hash")
+
+  printf '%s\n' "$expected_output"
```
I don't see a good reason to save the output in a variable and use it again in the `printf`. Why not use the printf directly?

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+
+# This function tests the self-update functionality of kworkflow for the master
+# branch.
+function test_self_update_master()
```
maybe use `test_self_update_from_master_branch`.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+# branch.
+function test_self_update_master()
+{
+  local kw_branch='master'
```
Since this function is dedicated to the master, why do you need this variable? You can use this constant direcly. Or if you still prefer to use the variable, at least use a name that makes clear that you are using the master branch.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+    expected_output=$(get_expected_output "$container" "$kw_cloned_container_path" "$kw_branch")
+
+    # Specific update command for master branch
+    container_exec "$container" 'yes | bash ./kw self-update > /dev/null'
```
Why you are using `.kw`? Fwiu, you should use the installed version, i.e., `kw self-update`.

aquilamacedo replied:
I use `bash ./kw self-update` because it's the best approach. It is more reliable than using the global version, as running `kw self-update` could **change** the installed version, making it different from the one used in the integration test.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+# unstable branch.
+function test_self_update_unstable()
+{
+  local kw_branch='unstable'
```
My comments from the previous tests also apply to this test.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+  done
+}
+
+function oneTimeTearDown()
```
This should be after the `oneTimeStartUp()` function.

rodrigosiqueira commented on tests/integration/self_update_test.sh:
```diff
@@ -0,0 +1,214 @@
+{
+  # Check if the path is safe to remove
+  if is_safe_path_to_remove "$HOST_TEMP_DIR"; then
+    rm -rf "$HOST_TEMP_DIR"
```
avoid short parameters.

rodrigosiqueira commented on tests/integration/utils.sh:
```diff
@@ -448,3 +448,17 @@ function container_inspect()
+  local container="$1"
+
+  container_exec "$container" 'mktemp --directory'
+
```
remove this extra line.

rodrigosiqueira commented on tests/integration/utils.sh:
```diff
@@ -448,3 +448,17 @@ function container_inspect()
+  container_exec "$container" 'mktemp --directory'
+
+  if [ "$?" -ne 0 ]; then
+    fail 'Failed to create a temporary directory inside the container'
```
If it fails, you must return an error code and check this function return in the test.

aquilamacedo commented:
changes made, could you give it a final review? @rodrigosiqueira

rodrigosiqueira commented:
PR merged into the unstable branch. Thanks
